net/port.go

```diff
@@ -0,0 +1,19 @@
+// Package net provides facilities for working with network connections.
+package net
+
+import (
+	"net"
+)
+
+// FreeTCPPort returns an unused local port.
+//
+// This is useful for tests that need to bind to a port without risking a conflict.
+func FreeTCPPort() (int, error) {
```

Comment:
There is a race condition with this approach. If you call this function to get a port and bind this port in another call, it might not be free anymore by that time. Ideally, you want to call `net.Listen("tcp", ":0")`, as you did, but use the socket that is already bound instead of closing it. For instance: https://github.com/OffchainLabs/nitro/blob/v2.4.0-rc.4/cmd/nitro/init_test.go#L120

Reply:
I strongly believe that the risk of a race is very low. The code won't race with other code which uses this function or binds to port 0. It will only race with code that hardcodes a port that happens to collide with this code.
See the additional inline comments in https://github.com/OffchainLabs/nitro/pull/2372/commits/04193dd15cedfb604e68e327ea0f5ff7ec539db2

Reply:
I think I'm missing something..
Why doesn't it race?
Assuming two threads implement logic like that:
```
port, err := FreeTCPPort()
newServer := attach(localhost, port)
```
What prevents the following race?
```
Thread A | open port 0 -> assign port 2000 -> port close
Thread B | open port 0 -> assign port 2000 -> port close
Thread A | attach port 2000 -> success
Thread B | attach port 2000 -> fail
```